master/buildbot/reporters/pushover.py

```diff
@@ -0,0 +1,294 @@
+            return False  # ignore this build
+        if self.tags is not None and \
+                not self.matchesAnyTag(builder['tags']):
+            return False  # ignore this build
+
+        if "change" in self.mode:
+            prev = build['prev_build']
+            if prev and prev['results'] != results:
+                return True
+        if "failing" in self.mode and results == FAILURE:
+            return True
+        if "passing" in self.mode and results == SUCCESS:
+            return True
+        if "problem" in self.mode and results == FAILURE:
+            prev = build['prev_build']
+            if prev and prev['results'] != FAILURE:
+                return True
+        if "warnings" in self.mode and results == WARNINGS:
+            return True
+        if "exception" in self.mode and results == EXCEPTION:
+            return True
+        if "cancelled" in self.mode and results == CANCELLED:
+            return True
+
+        return False
+
+    @defer.inlineCallbacks
+    def buildMessage(self, name, builds, results):
+        patches = []
+        logs = []
+        msg = {'message': ""}
+        users = set()
+        for build in builds:
+            if 'prev_build' in build and build['prev_build'] is not None:
+                previous_results = build['prev_build']['results']
+            else:
+                previous_results = None
+            blamelist = yield utils.getResponsibleUsersForBuild(self.master, build['buildid'])
+            msgdict = yield self.messageFormatter.formatMessageForBuildResults(
+                self.mode, name, build['buildset'], build, self.master,
+                previous_results, blamelist)
+            users.update(set(blamelist))
+            if msgdict['type'] != 'plain':
+                "'%s' message type must be 'plain'." % msgdict['type']
+            msg['message'] += msgdict['body']
+            try:
+                msg['priority'] = self.priorities[{
+                    CANCELLED: 'cancelled',
+                    EXCEPTION: 'exception',
+                    FAILURE: 'failing',
+                    SUCCESS: 'passing',
+                    WARNINGS: 'warnings'}[build['results']]]
+            except KeyError:
+                pass
+            if 'subject' in msgdict:
+                msg['title'] = msgdict['subject']
+            else:
+                msg['title'] = self.subject % {'result': Results[results],
+                                               'projectName': self.master.config.title,
+                                               'title': self.master.config.title,
+                                               'builder': name}
+
+        yield self.sendMessage(**msg)
+
+    def sendMessage(self, **kwargs):
+        twlog.msg("sending pushover notification")
+        client = Client(self.user_key, api_token=self.api_token)
+        kwargs.update(self.other_args)
+        return client.send_message(**kwargs)
```

Comment:
this api call is probably blocking. we dont allow blocking api inside the main thread.